### PR TITLE
plan, executor: no longer treat dml as logical/physical plan.

### PR DESCRIPTION
**Pull request by hanfei1991**

### CODE REVIEW

**Comment by shenli** [comment]
Why should we do this?

**Comment by coocood** [comment]
@hanfei1991 
Does it solve the expensive query log issue?

**Comment by hanfei1991** [comment]
@shenli make code cleaner, better and stronger.

**Comment by hanfei1991** [comment]
@coocood I don't understand what you mean.

File: executor/builder.go [hunk 1]
```diff
@@ -792,12 +790,12 @@ func (b *executorBuilder) buildUpdate(v *plan.Update) Executor {
...
 func (b *executorBuilder) buildDelete(v *plan.Delete) Executor {
 	tblID2table := make(map[int64]table.Table)
-	for id := range v.Schema().TblID2Handle {
+	for id := range v.SelectPlan.Schema().TblID2Handle {
 		tblID2table[id], _ = b.is.TableByID(id)
 	}
 	return &DeleteExec{
 		baseExecutor: newBaseExecutor(nil, b.ctx),
-		SelectExec:   b.build(v.Children()[0]),
+		SelectExec:   b.build(v.SelectPlan),
```
Comment (zz-jason) [inline executor/builder.go]: ditto

File: executor/builder.go [hunk 2]
```diff
@@ -778,12 +776,12 @@ func (b *executorBuilder) buildUnion(v *plan.Union) Executor {
...
 func (b *executorBuilder) buildUpdate(v *plan.Update) Executor {
 	tblID2table := make(map[int64]table.Table)
-	for id := range v.Schema().TblID2Handle {
+	for id := range v.SelectPlan.Schema().TblID2Handle {
 		tblID2table[id], _ = b.is.TableByID(id)
 	}
 	return &UpdateExec{
 		baseExecutor: newBaseExecutor(nil, b.ctx),
-		SelectExec:   b.build(v.Children()[0]),
+		SelectExec:   b.build(v.SelectPlan),
```
Comment (zz-jason) [inline executor/builder.go]: after building, we should check `b.err`:
```go
selectExec := b.build(v.SelectPlan)
if b.err != nil {
    b.err = errors.Trace(b.err)
    return nil
}
```

Comment (hanfei1991) [reply executor/builder.go]: if b.err is not nil, it's ok to return any executor.

Comment (zz-jason) [reply executor/builder.go]: It's better to discover error as soon as possible so that we can avoid useless executor building.

**Comment by shenli** [comment]
@hanfei1991 Please fix the conflicts.

File: plan/logical_plan_builder.go [hunk 3]
```diff
@@ -1827,8 +1827,8 @@ func (b *planBuilder) buildUpdate(update *ast.UpdateStmt) LogicalPlan {
 		OrderedList: orderedList,
 		IgnoreErr:   update.IgnoreErr,
 	}.init(b.ctx)
-	setParentAndChildren(updt, p)
 	updt.SetSchema(p.Schema())
+	updt.SelectPlan, b.err = doOptimize(b.optFlag, p, b.ctx)
```
Comment (XuHuaiyu) [inline plan/logical_plan_builder.go]: check b.err

File: plan/plan.go [hunk 4]
```diff
@@ -52,6 +52,9 @@ type Plan interface {
 	// findColumn finds the column in basePlan's schema.
 	// If the column is not in the schema, returns error.
 	findColumn(*ast.ColumnName) (*expression.Column, int, error)
+
+	// ResolveIndices resolves the indices for columns. After doing this, the columns can evaluate the rows by their indices.
+	ResolveIndices()
```
Comment (XuHuaiyu) [inline plan/plan.go]: why move this.

File: plan/resolve_indices.go [hunk 5]
```diff
@@ -216,9 +216,11 @@ func (p *Insert) ResolveIndices() {
 
 // ResolveIndices implements Plan interface.
 func (p *basePlan) ResolveIndices() {
-	for _, cols := range p.schema.TblID2Handle {
-		for _, col := range cols {
-			col.ResolveIndices(p.schema)
+	if p.schema != nil {
```
Comment (XuHuaiyu) [inline plan/resolve_indices.go]: comment for when this check will be true.

File: plan/plan.go (hunk 4 above)
Comment (hanfei1991) [reply plan/plan.go]: Insert/Delete/Update also needs resolveIndices

File: plan/logical_plan_builder.go (hunk 3 above)
Comment (hanfei1991) [reply plan/logical_plan_builder.go]: If b.err is not nil, we can also return. It's no need to check this.

File: plan/resolve_indices.go (hunk 5 above)
Comment (hanfei1991) [reply plan/resolve_indices.go]: I also don't know when it would fail. It's strange and I will refactor the schema setting in the future.

**Comment by hanfei1991** [comment]
@zz-jason @XuHuaiyu PTAL

File: executor/builder.go [hunk 6]
```diff
@@ -319,8 +319,9 @@ func (b *executorBuilder) buildInsert(v *plan.Insert) Executor {
 		GenExprs:              v.GenCols.Exprs,
 		needFillDefaultValues: v.NeedFillDefaultValue,
 	}
-	if len(v.Children()) > 0 {
-		ivs.SelectExec = b.build(v.Children()[0])
+	ivs.SelectExec = b.build(v.SelectPlan)
+	if b.err != nil {
```
Comment (zz-jason) [inline executor/builder.go]: b.err = errors.Trace(b.err)

File: executor/builder.go [hunk 7]
```diff
@@ -795,12 +796,16 @@ func (b *executorBuilder) buildUnion(v *plan.Union) Executor {
 
 func (b *executorBuilder) buildUpdate(v *plan.Update) Executor {
 	tblID2table := make(map[int64]table.Table)
-	for id := range v.Schema().TblID2Handle {
+	for id := range v.SelectPlan.Schema().TblID2Handle {
 		tblID2table[id], _ = b.is.TableByID(id)
 	}
+	selExec := b.build(v.SelectPlan)
+	if b.err != nil {
```
Comment (zz-jason) [inline executor/builder.go]: ditto

**Comment by hanfei1991** [comment]
@XuHuaiyu PTAL

File: plan/logical_plan_test.go [hunk 8]
```diff
@@ -612,19 +612,19 @@ func (s *testPlanSuite) TestPlanBuilder(c *C) {
 		},
 		{
 			sql:  "update t set t.a = t.a * 1.5 where t.a >= 1000 order by t.a desc limit 10",
-			plan: "DataScan(t)->Sel([ge(test.t.a, 1000)])->Sort->Limit->*plan.Update",
+			plan: "TableReader(Table(t)->Limit)->Limit->Update",
```
Comment (shenli) [inline plan/logical_plan_test.go]: Why DataScan is replaced with TableReader?

**Comment by shenli** [comment]
Should you consider the `Replace` statement in this PR?

**Comment by hanfei1991** [comment]
@shenli replace is a kind of insert

File: plan/logical_plan_test.go (hunk 8 above)
Comment (hanfei1991) [reply plan/logical_plan_test.go]: We build update's `SelectPlan` at one time.

**Comment by shenli** [comment]
@hanfei1991 CI fail.

**Pull request merged**